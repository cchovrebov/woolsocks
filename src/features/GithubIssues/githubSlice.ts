/* eslint-disable max-lines */
import { createSlice } from '@reduxjs/toolkit'

import { githubApi } from 'common/api'

export interface Organization {
  avatar_url: string
  description: string
  events_url: string
  hooks_url: string
  id: number
  issues_url: string
  login: string
  members_url: string
  node_id: string
  public_members_url: string
  repos_url: string
  url: string
}

export interface GithubState {
  isLoading: boolean
  errorMessage: string
  accessToken: string
  organizations: Organization[] | null
  repositories: any
  organization: string
  repository: string
  issues: any[] | null
}

export const initialState: GithubState = {
  isLoading: false,
  errorMessage: '',
  accessToken: '',
  organizations: null,
  organization: '',
  repositories: null,
  repository: '',
  issues: undefined,
}

export const githubSlice = createSlice({
  name: 'githubSlice',
  initialState,
  reducers: {
    updateError: (state, action) => {
      state.errorMessage = action.payload
    },
    updateFormValues: (state, action) => {
      state[action.payload.field] = action.payload.value
    },
    updateOrganizations: (state, action) => {
      state.organizations = action.payload
    },
    updateRepositories: (state, action) => {
      state.repositories = action.payload
    },
    updateIssues: (state, action) => {
      state.issues = action.payload
    },
    updateLoading: (state, action) => {
      state.isLoading = action.payload
    },
  },
})

export const {
  updateFormValues,
  updateError,
  updateOrganizations,
  updateRepositories,
  updateIssues,
  updateLoading,
} = githubSlice.actions

export function fetchGitRepos(data: {
  accessToken: string
  organization: string
}) {
  return githubApi.get<void>(`/orgs/${data.organization}/repos`, {
    headers: {
      Authorization: `token ${data.accessToken}`,
      Accept: 'application/vnd.github.v3+json',
    },
  })
}

export function fetchGitOrg(data: { accessToken: string }) {
  return githubApi.get<void>('/user/orgs', {
    headers: {
      Authorization: `token ${data.accessToken}`,
      Accept: 'application/vnd.github.v3+json',
    },
  })
}

export function fetchIssuesByRepo(data: {
  accessToken: string
  repository: string
  owner: string
}) {
  return githubApi.get<void>(`/repos/${data.owner}/${data.repository}/issues`, {
    headers: {
      Authorization: `token ${data.accessToken}`,
      Accept: 'application/vnd.github.v3+json',
    },
  })
}

export default githubSlice.reducer
