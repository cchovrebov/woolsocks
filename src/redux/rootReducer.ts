import githubReducer from 'features/GithubIssues/githubSlice'

const reducers = {
  reducer: {
    githubReducer,
  },
}

export default reducers
