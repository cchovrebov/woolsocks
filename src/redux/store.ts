import { Action, configureStore, ThunkAction } from '@reduxjs/toolkit'

import reducers from './rootReducer'

export const store = configureStore({
  ...reducers,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: true,
      immutableCheck: true,
    }),
})

export type AppDispatch = typeof store.dispatch
export type RootState = ReturnType<typeof store.getState>
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>
