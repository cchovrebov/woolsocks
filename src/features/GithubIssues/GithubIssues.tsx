import React, { useEffect } from 'react'
import CloseIcon from '@mui/icons-material/Close'
import { Alert, Collapse, IconButton, LinearProgress } from '@mui/material'
import DataTable from 'components/DataTable/DataTable'
import GithubForm from 'components/GithubForm/GithubForm'
import { find, map } from 'lodash'

import { useAppDispatch, useAppSelector } from 'common/hooks/redux'

import {
  fetchGitOrg,
  fetchGitRepos,
  fetchIssuesByRepo,
  GithubState,
  updateError,
  updateFormValues,
  updateIssues,
  updateLoading,
  updateOrganizations,
  updateRepositories,
} from './githubSlice'

interface IssuesData {
  title: string
  comment: string
  state: string
  user: string
}

const issuesColumns: IssuesData = {
  title: 'title',
  comment: 'comment',
  state: 'state',
  user: 'user',
}

const GithubIssues = () => {
  const dispatch = useAppDispatch()
  const {
    isLoading,
    accessToken,
    organizations,
    organization,
    repositories,
    repository,
    issues,
    errorMessage,
  } = useAppSelector<GithubState>((state) => state.githubReducer)

  useEffect(() => {
    if (organization && !repositories?.length) {
      dispatch(updateLoading(true))
      fetchGitRepos({ accessToken, organization })
        .then((res) => {
          dispatch(updateRepositories(res.data))
          dispatch(updateLoading(false))
          dispatch(updateError(''))
        })
        .catch((err) => {
          dispatch(updateError(err.message))
          dispatch(updateLoading(false))
        })
    }
    if (repository) {
      const repoObt = find(repositories, { name: repository })
      if (repoObt?.owner?.login) {
        dispatch(updateLoading(true))
        fetchIssuesByRepo({
          accessToken,
          repository,
          owner: repoObt?.owner?.login,
        })
          .then((res) => {
            dispatch(updateIssues(res.data))
            dispatch(updateLoading(false))
            dispatch(updateError(''))
          })
          .catch((err) => {
            dispatch(updateError(err.message))
            dispatch(updateLoading(false))
          })
      }
    }
  }, [dispatch, organization, repository, accessToken, repositories])

  const onFormChange = (value: string, field: string) => {
    dispatch(updateFormValues({ value, field }))
  }

  const onSubmit = () => {
    if (!accessToken) {
      dispatch(updateError('Access token field is required'))
    } else {
      dispatch(updateLoading(true))
      fetchGitOrg({ accessToken })
        .then((res) => {
          dispatch(updateOrganizations(res.data))
          dispatch(updateLoading(false))
          dispatch(updateError(''))
        })
        .catch(() => {
          dispatch(updateError('Bad credentials'))
          dispatch(updateLoading(false))
        })
    }
  }

  const formatIssuesData = (rawData: any[]): IssuesData[] => {
    return map(rawData, (item: any) => ({
      title: item.title,
      comment: item.body,
      state: item.state,
      user: item.user.login,
    }))
  }

  return (
    <>
      {isLoading && <LinearProgress />}
      <Collapse in={!!errorMessage}>
        <Alert
          action={
            <IconButton
              aria-label="close"
              color="inherit"
              size="small"
              onClick={() => dispatch(updateError(''))}
            >
              <CloseIcon fontSize="inherit" />
            </IconButton>
          }
          sx={{ mb: 2 }}
          style={{ margin: '1rem 0' }}
          variant="filled"
          severity="error"
        >
          {errorMessage}
        </Alert>
      </Collapse>
      <GithubForm
        onChange={onFormChange}
        onSubmit={onSubmit}
        accessToken={accessToken}
        organizations={organizations}
        organization={organization}
        repositories={repositories}
        repository={repository}
      />
      {issues && (
        <DataTable data={formatIssuesData(issues)} columns={issuesColumns} />
      )}
    </>
  )
}

export default GithubIssues
