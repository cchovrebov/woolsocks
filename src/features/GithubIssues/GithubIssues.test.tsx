import React from 'react'
import { Provider } from 'react-redux'
import { render, screen } from '@testing-library/react'
import { store } from 'redux/store'

import GithubIssues from './GithubIssues'

const createElement = () => {
  return render(
    <Provider store={store}>
      <GithubIssues />
    </Provider>
  )
}

describe('GithubIssues Recently created', () => {
  it('Renders correctly', async () => {
    const { asFragment } = createElement()
    expect(asFragment()).toMatchSnapshot()
  })

  it('Github form renders', async () => {
    createElement()
    const githubForm = screen.getByTestId('github-form')
    expect(githubForm).toMatchSnapshot()
    expect(!!githubForm).toBeTruthy()
  })
})
