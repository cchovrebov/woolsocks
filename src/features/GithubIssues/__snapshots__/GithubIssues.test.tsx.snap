// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubIssues Recently created Github form renders 1`] = `
<form
  autocomplete="off"
  class="github-form MuiBox-root css-0"
  data-testid="github-form"
  novalidate=""
>
  <div
    class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for="component-error"
    >
      Access token
    </label>
    <div
      class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl github-form-accessToken css-1480iag-MuiInputBase-root-MuiInput-root"
      data-testid="github-form-accessToken"
    >
      <input
        aria-describedby="component-error-text"
        aria-invalid="false"
        class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        type="password"
        value=""
      />
    </div>
  </div>
  <button
    class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
    data-testid="github-form-button"
    tabindex="0"
    type="button"
  >
    Fetch organizations
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</form>
`;

exports[`GithubIssues Recently created Renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
    style="min-height: 0px;"
  >
    <div
      class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
    >
      <div
        class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-filledError MuiAlert-filled css-1a1lkbh-MuiPaper-root-MuiAlert-root"
          role="alert"
          style="margin: 1rem 0px;"
        >
          <div
            class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="ErrorOutlineIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
          </div>
          <div
            class="MuiAlert-message css-acap47-MuiAlert-message"
          />
          <div
            class="MuiAlert-action css-ki1hdl-MuiAlert-action"
          >
            <button
              aria-label="close"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall css-1e0d89p-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                data-testid="CloseIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <form
    autocomplete="off"
    class="github-form MuiBox-root css-0"
    data-testid="github-form"
    novalidate=""
  >
    <div
      class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="component-error"
      >
        Access token
      </label>
      <div
        class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl github-form-accessToken css-1480iag-MuiInputBase-root-MuiInput-root"
        data-testid="github-form-accessToken"
      >
        <input
          aria-describedby="component-error-text"
          aria-invalid="false"
          class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="github-form-button"
      tabindex="0"
      type="button"
    >
      Fetch organizations
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </form>
</DocumentFragment>
`;
